src/Common/test/Certificates.Test/ConfigureCertificateOptionsTest.cs

```diff
@@ -17,176 +18,287 @@ public sealed class ConfigureCertificateOptionsTest
+
+            fileProvider.ReplaceFile(MemoryFileProvider.DefaultAppSettingsFileName,
+                BuildAppSettingsJson(certificateName, secondCertificateFilePath, secondPrivateKeyFilePath));
+
+            fileProvider.NotifyChanged();
+
+            configurationChangeCalled.Should().BeTrue("file path information changed");
+            optionsMonitor.Get(certificateName).Certificate.Should().Be(secondX509);
+        }
+
+        bool certificateContentsChangeCalled = false;
+        bool keyContentsChangeCalled = false;
+
+        List<FilePathInOptionsChangeTokenSource<CertificateOptions>> matchingTokenSources =
+        [
+            .. tokenSources.OfType<FilePathInOptionsChangeTokenSource<CertificateOptions>>()
+        ];
+
+        matchingTokenSources.Should().HaveCount(2);
+
+        List<IChangeToken> changeTokens = [.. matchingTokenSources.Select(s => s.GetChangeToken())];
+
+        using IDisposable certificateContentChangeToken = changeTokens[0].RegisterChangeCallback(_ => certificateContentsChangeCalled = true, null);
+        using IDisposable keyContentChangeToken = changeTokens[1].RegisterChangeCallback(_ => keyContentsChangeCalled = true, null);
+
+        certificateContentsChangeCalled.Should().BeFalse("certificate file content has not changed yet");
+        keyContentsChangeCalled.Should().BeFalse("key file content has not changed yet");
+        await File.WriteAllTextAsync(secondCertificateFilePath, firstInstanceCertificate, TestContext.Current.CancellationToken);
+        await File.WriteAllTextAsync(secondPrivateKeyFilePath, firstInstancePrivateKey, TestContext.Current.CancellationToken);
+
+        using Task pollTask = PollCertificateAsync(optionsMonitor, certificateName, firstX509, TestContext.Current.CancellationToken);
+        await pollTask.WaitAsync(TimeSpan.FromSeconds(4), TestContext.Current.CancellationToken);
+
+        certificateContentsChangeCalled.Should().BeTrue("certificate file contents changed");
+        keyContentsChangeCalled.Should().BeTrue("key file contents changed");
+        optionsMonitor.Get(certificateName).Certificate.Should().Be(firstX509);
+    }
+
+    private static string BuildAppSettingsJson(string certName, string certPath, string keyPath)
+    {
+        string certificateBlock = $"""
+                "CertificateFilePath": {JsonSerializer.Serialize(certPath)},
+                "PrivateKeyFilePath": {JsonSerializer.Serialize(keyPath)}
+            """;
+
+        string namedCertificateSection = string.IsNullOrEmpty(certName) ? certificateBlock : $"{JsonSerializer.Serialize(certName)}: {{ {certificateBlock} }}";
+
+        return $$"""
+            {
+              "Certificates": {
+                {{namedCertificateSection}}
+              }
+            }
+            """;
+    }
+
+    private static async Task PollCertificateAsync(IOptionsMonitor<CertificateOptions> optionsMonitor, string certificateName,
```

Comment:
Suggested name: `WaitUntilCertificateChangedTo`, with `expectedCertificate` as the first parameter.